Fix part of #4057: Increase test coverage to 100% for HintObjectFactory.js and InteractionObjectFactory.js.

YashJipkate opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR fixes part of #4057 by adding tests for HintObjectFactory.js and InteractionObjectFactory.js

![IMG_20190311_005044](https://user-images.githubusercontent.com/34203227/54090232-f3bec800-4397-11e9-8b9c-0155e54309ed.png)

<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

YashJipkate committed:
Increase coverage for two files

kevinlee12 commented on core/templates/dev/head/domain/exploration/HintObjectFactorySpec.js:
```diff
@@ -0,0 +1,35 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
```
Year :)

YashJipkate replied:
Oh missed that, Done!

kevinlee12 commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,233 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
```
Ditto

YashJipkate replied:
Done.

kevinlee12 commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,233 @@
+  var solutionDict = null;
+  var hintsDict = null;
+  var interactionDict = null;
+  beforeEach(module('oppia'));
```
Add space above this

YashJipkate replied:
Done.

kevinlee12 commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,233 @@
+  var hintsDict = null;
+  var interactionDict = null;
+  beforeEach(module('oppia'));
+  beforeEach(inject(function($injector) {
```
Ditto on spacing

YashJipkate replied:
Done.

kevinlee12 commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,233 @@
+    testInteraction.setHints([newHint]);
+    expect(testInteraction.hints).toEqual([newHint]);
+  });
+  it('should correctly copy from other interaction', function() {
```
Add space above

YashJipkate replied:
Done.

kevinlee12 reviewed (commented):
Hi @YashJipkate , I left some sanity comments, ptal!

YashJipkate committed:
Address Review

YashJipkate commented:
Hi, @kevinlee12 @DubeySandeep @apb7 @bansalnitish Could anyone of you PTAL?
Thanks!

kevinlee12 reviewed (approved):
lgtm! 

passing it over to @DubeySandeep for second pass

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+  beforeEach(module('oppia'));
+
+  beforeEach(inject(function($injector) {
+    iof = $injector.get('InteractionObjectFactory');
```
Undefined iof.

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+
+  beforeEach(inject(function($injector) {
+    iof = $injector.get('InteractionObjectFactory');
+    oof = $injector.get('OutcomeObjectFactory');
```
Ditto. (same for next 3 lines)

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+        refresher_exploration_id: null,
+        missing_prerequisite_skill_id: null
+      },
+      training_data: 'training_data',
```
Are you sure that training data is a string?

YashJipkate replied:
Ah ok, I forgot to test the actual type. It is actually an array as can be seen [here](https://github.com/oppia/oppia/blob/e926756802fb70f6a6b22808663337ea645a6b6d/core/templates/dev/head/pages/exploration_player/AnswerClassificationServiceSpec.js#L486)

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+        missing_prerequisite_skill_id: null
+      },
+      training_data: 'training_data',
+      tagged_misconception_id: 'tagged_misconception_id'
```
Are you sure that tagged_misconception_id is a string?

YashJipkate replied:
Done. Found the correct type [here](https://github.com/oppia/oppia/blob/develop/core/domain/state_domain.py#L81)

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+    agof = $injector.get('AnswerGroupObjectFactory');
+    hof = $injector.get('HintObjectFactory');
+    sof = $injector.get('SolutionObjectFactory');
+    defaultOutcome = {
```
Make it defaultOutcomeDict.

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+    sof = $injector.get('SolutionObjectFactory');
+    defaultOutcome = {
+      dest: 'dest_default',
+      feedback: 'feedback',
```
are you sure that the feedback is a string?

YashJipkate replied:
Done. Found the correct type [here](https://github.com/oppia/oppia/blob/e926756802fb70f6a6b22808663337ea645a6b6d/core/templates/dev/head/domain/state_card/StateCardObjectFactorySpec.js#L50)

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+      refresher_exploration_id: null,
+      missing_prerequisite_skill_id: null
+    };
+    answerGroups = [{
```
answerGroupsDict*

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+    interactionDict = {
+      answer_groups: answerGroups,
+      confirmed_unclassified_answers: [],
+      customization_args: [],
```
are you sure that it's a list?

YashJipkate replied:
Done. Found the correct type [here](https://github.com/oppia/oppia/blob/e926756802fb70f6a6b22808663337ea645a6b6d/core/templates/dev/head/pages/exploration_player/ImagePreloaderServiceSpec.js#L92).

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+      rule_specs: [],
+      outcome: {
+        dest: 'dest_1',
+        feedback: 'feedback',
```
Ditto.

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+      rule_specs: [],
+      outcome: {
+        dest: 'dest_1',
+        feedback: 'feedback',
```
Is this a string type?

YashJipkate replied:
Corrected. Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+  it('should correctly set the new default outcome', function() {
+    var newDefaultOutcomeDict = {
+      dest: 'dest_default_new',
+      feedback: 'feedback_new',
```
Ditto

YashJipkate replied:
Corrected. Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+    };
+    newDefaultOutcome = oof.createFromBackendDict(newDefaultOutcomeDict);
+    expect(testInteraction.defaultOutcome).toEqual(
+      oof.createFromBackendDict(defaultOutcome));
```
can we validate the dict as you have done above?

YashJipkate replied:
Sure. Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+  });
+
+  it('should correctly set the new customization args', function() {
+    var newCustomizationArgs = ['arg_1_new', 'arg_2_new'];
```
are you sure it's a list?

YashJipkate replied:
Corrected. Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+    };
+    newSolution = sof.createFromBackendDict(newSolutionDict);
+    expect(testInteraction.solution).toEqual(
+      sof.createFromBackendDict(solutionDict));
```
Ditoo (As above)

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,236 @@
+  });
+
+  it('should correctly set the new hint', function() {
+    var newHintsDict = {
```
newHintDict*

YashJipkate replied:
Done.

DubeySandeep reviewed (changes requested):
@YashJipkate, I've left a few comments PTAL!

 Also, I'm not sure whether we should write tests for all of the behaviours in the `HintObjectFactorySpec.js` or just which aren't covered. I'll let you know after consulting others. Thanks!

DubeySandeep commented:
@YashJipkate, I asked others for help regarding  
> Also, I'm not sure whether we should write tests for all of the behaviors in the HintObjectFactorySpec.js or just which aren't covered. I'll let you know after consulting others. Thanks!

Here's what I got: Ideally we should cover the test for the entire file in its own test file.

YashJipkate committed:
Address review

YashJipkate commented:
Hi, @DubeySandeep, Thanks for the review. I have addressed the review comments. PTAL!

Sorry for all the type mess, I think I missed them or incorrectly saw them. Will take care next time.

Regarding your concern on `HintObjectFactorySpec.js`, I am not sure I understood you. The only functions that are not tested are the ones that are used to create the instance. However, these functions are covered while carrying out the tests of the other functions thereby giving the 100% coverage. So I don't think that making tests for these functions is required.
Is that what you required?

Thanks.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,291 @@
+    };
+    var otherInteraction = iof.createFromBackendDict(otherInteractionDict);
+    testInteraction.copy(otherInteraction);
+    expect(testInteraction).toEqual(otherInteraction);
```
@YashJipkate, can we test: changing the otherInteraction values doesn't get reflected to testInteraction?
(we are trying to test the behavior of the function although the test coverage is 100%)

(I had added this comment in my last review I don't know why it wasn't that.)

YashJipkate replied:
I have added a test for this. PTAL! Thanks

DubeySandeep commented:
> Regarding your concern on HintObjectFactorySpec.js, I am not sure I understood you. The only functions that are not tested are the ones that are used to create the instance. However, these functions are covered while carrying out the tests of the other functions thereby giving the 100% coverage. So I don't think that making tests for these functions is required.
Is that what you required?

Yep, I was asking you to write the test for all the functionality/method in HintObjectFactory, let's not depend on other part of test covering few functionality.

DubeySandeep commented:
Also, can you please update the PR title with the file name instead of "two files"?

YashJipkate committed:
Added additional test for copy functionality

YashJipkate committed:
Added additional tests

YashJipkate commented:
Hi @DubeySandeep I have added the tests as you directed PTAL! Thanks.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,310 @@
+      id: 'interaction_id_new_changed',
+      solution: newSolutionDict
+    };
+    otherInteraction = iof.createFromBackendDict(changedOtherInteractionDict);
```
@YashJipkate, I think a better way to test the copy function would be to change the internal properties of the otherInteraction instead of assigning a new value to it. 

maybe changing the customization_args, adding a hint, changing the internal property of solution etc.

DubeySandeep commented on core/templates/dev/head/domain/exploration/HintObjectFactorySpec.js:
```diff
@@ -0,0 +1,56 @@
+    );
+  }));
+
+  it('should create a hint object from a dict correctly', inject(function() {
```
Can we have an integration test that will check "to and from backendDict" in a single test?

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/HintObjectFactorySpec.js:
```diff
@@ -0,0 +1,56 @@
+// limitations under the License.
+
+/**
+ * @fileoverview unit tests for hint object factory.
```
Unit*

YashJipkate replied:
Done.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,310 @@
+        missing_prerequisite_skill_id: null
+      },
+      training_data: ['training_data'],
+      tagged_misconception_id: 'tagged_misconception_id'
```
are you sure this is a string?

YashJipkate replied:
Oh missed it again. Sorry. Done. 
Thanks.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,310 @@
+      id: 'interaction_id_new_changed',
+      solution: newSolutionDict
+    };
+    otherInteraction = iof.createFromBackendDict(changedOtherInteractionDict);
```
@YashJipkate, I think a better way to test the copy function would be to change the internal properties of the otherInteraction instead of assigning a new value to it. 

maybe changing the customization_args, adding a hint, changing the internal property of solution etc.

YashJipkate replied:
I have changed the customization_args. PTAL!
Thanks.

DubeySandeep reviewed (commented):
@YashJipkate, This looks great now! I've left the last few comments once they are resolved I think then we can merge this PR.

Thanks!

YashJipkate committed:
Review address

YashJipkate committed:
Removed extra unnecessary test

YashJipkate commented:
@DubeySandeep PTAL! Thanks.

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,305 @@
+    var otherInteraction = iof.createFromBackendDict(otherInteractionDict);
+    testInteraction.copy(otherInteraction);
+    expect(testInteraction).toEqual(otherInteraction);
+    otherInteraction.setCustomizationArgs({
```
You are still setting up a different value, try getCustomizationArgs and change the `customArgNew` value.

(maybe we can also change a few more values?)

YashJipkate replied:
Hi @DubeySandeep, there is no `getCustomizationArgs` function in InteractionObjectFactory. The only way to access the `customArgs` is by converting the interaction object to dict and getting the value from there. So should I get the `customArgs` by this way and then change it to `customArgNew`?

YashJipkate replied:
Hi, @DubeySandeep made some changes, PTAL!
Thanks

YashJipkate committed:
Address review

YashJipkate commented:
Hi @DubeySandeep PTAL! Thanks

DubeySandeep commented on core/templates/dev/head/domain/exploration/InteractionObjectFactorySpec.js:
```diff
@@ -0,0 +1,307 @@
+    expect(testInteraction).toEqual(otherInteraction);
+    otherInteractionDict = otherInteraction.toBackendDict();
+    var customizationArgsNewDict = otherInteractionDict.customization_args;
+    customizationArgsNewDict = {
```
@YashJipkate, You aren't changing the internal property of the dict, you are just replacing with a new one.

Maybe you can try this `otherInteraction.customizationArgs.customArgNew.value = "something_new"`.

(we don't have a getter so I think we can access customizationArgs this way. See StateObjectFactory directly access this value.)

DubeySandeep replied:
Wait, I realized that initially we have customizationArgs as a list and now we are making it dict?

YashJipkate replied:
Oh, it's a dict only, as pointed [here](https://github.com/oppia/oppia/pull/6428#discussion_r265467528). Corrected it.

Also made the change as you said.
PTAL! Thanks!

YashJipkate committed:
Better testing tactic

DubeySandeep reviewed (approved):
LGTM!

YashJipkate commented:
Hi @DubeySandeep, could this be merged now since it's approved?
Thanks!
